File: app/src/main/java/com/zibete/proyecto1/Adapters/ChatDiffCallback.java
package com.zibete.proyecto1.Adapters;

import android.os.Bundle;

import androidx.annotation.Nullable;
import androidx.recyclerview.widget.DiffUtil;

import com.zibete.proyecto1.POJOS.ChatWith;

import java.util.ArrayList;
import java.util.List;

public class ChatDiffCallback extends DiffUtil.Callback {

    private ArrayList<ChatWith> mNewList;
    private ArrayList<ChatWith> mOldList;


    public ChatDiffCallback(List<ChatWith> newlist, List<ChatWith> oldlist) {
        this.mNewList = (ArrayList<ChatWith>) newlist;
        this.mOldList = (ArrayList<ChatWith>) oldlist;
    }

    @Override
    public int getOldListSize() {
        return mOldList.size();
    }

    @Override
    public int getNewListSize() {
        return mNewList.size();
    }

    @Override
    public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
        return mOldList.get(oldItemPosition).getwUserID()==mNewList.get(newItemPosition).getwUserID();
    }

    @Override
    public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
        return mOldList.get(oldItemPosition).equals(mNewList.get(newItemPosition));
    }

    @Nullable
    @Override
    public Object getChangePayload(int oldItemPosition, int newItemPosition) {
        ChatWith newItem = mNewList.get(newItemPosition);
        ChatWith oldItem = mOldList.get(oldItemPosition);

        Bundle diff = new Bundle();


        if(!newItem.getwUserID().equals(oldItem.getwUserID())){
            diff.putString("id", newItem.getwUserID());
        }


        else{




            if(!newItem.getNoVisto().equals(oldItem.getNoVisto())){
                diff.putString("date", newItem.getwDate());
            }


            diff.putString("other", newItem.getwUserID());





        }



        if (diff.size()==0){
            return null;
        }
        return diff;
    }
}

File: app/src/main/java/com/zibete/proyecto1/adapters/ChatDiffCallback.kt
package com.zibete.proyecto1.adapters

import android.os.Bundle
import androidx.recyclerview.widget.DiffUtil
import com.zibete.proyecto1.model.Conversation

class ChatDiffCallback(
    private val newList: List<Conversation>,
    private val oldList: List<Conversation>
) : DiffUtil.Callback() {

    override fun getOldListSize() = oldList.size

    override fun getNewListSize() = newList.size

    override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean =
        oldList[oldItemPosition].otherId === newList[newItemPosition].otherId

    override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean =
         oldList[oldItemPosition] == newList[newItemPosition]

    override fun getChangePayload(oldPos: Int, newPos: Int): Any? {
        val oldItem = oldList[oldPos]
        val newItem = newList[newPos]

        return Bundle().apply {
            if (oldItem.otherId != newItem.otherId)
                putString("id", newItem.otherId)

            if (oldItem.unreadCount != newItem.unreadCount)
                putString("date", newItem.date.toString())
        }.takeIf { !it.isEmpty }
    }

}



